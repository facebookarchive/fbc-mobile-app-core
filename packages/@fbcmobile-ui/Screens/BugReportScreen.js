/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import type {NavigationNavigatorProps} from 'react-navigation';
import type {NavigationScreenConfig} from 'react-navigation';

import Button from '@fbcmobile/ui/Components/Core/Button';
import FormInput from '@fbcmobile/ui/Components/FormInput/FormInput';
import NavigationService from '@fbcmobile/ui/Services/NavigationService';
import React, {useEffect, useRef, useState} from 'react';
import Spinner from 'react-native-spinkit';
import TextFormInput from '@fbcmobile/ui/Components/FormInput/TextFormInput';
import UserActionLogger from '@fbcmobile/ui/Logging/UserActionLogger';
import {Colors} from '@fbcmobile/ui/Theme';
import {EVENT} from '@fbcmobile/ui/Logging/UserActionEvents';
import {Icon} from 'react-native-material-ui';
import {Image, ScrollView, View} from 'react-native';
import {TextInput} from 'react-native';
import {fbt} from 'fbt';

type Props = NavigationNavigatorProps<{}, {params: {photoData: string}}>;

type ImageSize = {|
  width: number,
  height: number,
|};

const BugReportScreen = (props: Props) => {
  const {navigation} = props;
  const photoData = navigation.getParam('photoData');
  const [imageSize, setImageSize] = useState<?ImageSize>(null);
  const [details, setDetails] = useState<?string>(null);
  const [uploading, setUploading] = useState<boolean>(false);
  const inputRef = useRef<?React$ElementRef<typeof TextInput>>(null);

  useEffect(() => {
    Image.getSize(photoData, (width: number, height: number) =>
      setImageSize({width, height}),
    );
  }, [photoData]);

  return (
    <View style={styles.root}>
      <ScrollView>
        <View style={styles.container}>
          <FormInput
            title={fbt('Details', 'Text box header').toString()}
            description={fbt(
              "Please describe the bug you've observed",
              'Instructions for describing the observed bug',
            ).toString()}
            onPress={() => {
              inputRef.current && inputRef.current.focus();
            }}>
            <TextFormInput
              multiline={true}
              autoFocus={true}
              numberOfLines={3}
              value={details}
              onChangeText={text => setDetails(text)}
              ref={inputRef}
            />
          </FormInput>
          <FormInput title={fbt('Screenshot', 'Screenshot header').toString()}>
            {imageSize && (
              <Image
                style={{
                  ...styles.photoContainer,
                  width: imageSize.width / 4,
                  height: imageSize.width / 4,
                }}
                resizeMode="contain"
                source={{uri: photoData}}
              />
            )}
            <Button
              disabled={uploading}
              textAlign="left"
              rightIcon={
                uploading ? (
                  <Spinner
                    size={20}
                    color={Colors.DisabledText}
                    type="ThreeBounce"
                  />
                ) : (
                  <Icon
                    name="arrow-right"
                    iconSet="MaterialCommunityIcons"
                    color={Colors.White}
                    size={20}
                  />
                )
              }
              onPress={() => {
                setUploading(true);
                UserActionLogger.logBug({
                  key: EVENT.BUG_REPORT,
                  details: details,
                  screenShot: photoData,
                });
                setUploading(false);
                NavigationService.alert(
                  'success',
                  fbt('Bug Report Submitted', 'Label for Bug Report Submitted'),
                  fbt(
                    'Thank you for submitting a bug report!  Our team will look into it as soon as possible.',
                    'Text thanking the user for submitting a bug report',
                  ),
                );
                NavigationService.pop();
              }}>
              <fbt desc="Submit bug button text">Submit</fbt>
            </Button>
          </FormInput>
        </View>
      </ScrollView>
    </View>
  );
};

const styles = {
  root: {
    backgroundColor: Colors.BackgroundWhite,
  },
  container: {
    padding: 20,
  },
  photoContainer: {
    borderRadius: 10,
    marginBottom: 30,
    backgroundColor: Colors.Gray10,
    justifyContent: 'center',
    alignItems: 'center',
  },
};

const options: NavigationScreenConfig<*> = {
  headerShown: true,
  headerTitle: fbt('Bug Report', 'Bug Report screen title').toString(),
};

BugReportScreen.navigationOptions = options;

export default BugReportScreen;
