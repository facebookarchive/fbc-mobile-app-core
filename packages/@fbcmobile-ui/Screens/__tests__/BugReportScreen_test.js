/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import BugReportScreen from '../BugReportScreen';
import React from 'react';
import renderer from 'react-test-renderer';

describe('BugReportScreen Tests', () => {
  const navigation = {
    getParam: jest.fn(),
    setParams: jest.fn(),
    navigate: jest.fn(),
    isFocused: jest.fn(),
    addListener: jest.fn(),
    dismiss: jest.fn(),
    dispatch: jest.fn(),
    goBack: jest.fn(),
    state: {
      params: {
        photoData: '/media/pic.png',
      },
    },
    dangerouslyGetParent: jest.fn(),
  };

  test('BugReportScreen renders correctly', () => {
    const tree = renderer
      .create(<BugReportScreen navigation={navigation} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
