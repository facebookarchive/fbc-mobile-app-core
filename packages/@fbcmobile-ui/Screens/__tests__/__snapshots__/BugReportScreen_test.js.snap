// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BugReportScreen Tests BugReportScreen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FBFBFB",
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "padding": 20,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                },
                null,
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flex": 1,
                    "marginBottom": 10,
                  },
                  null,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#484848",
                      "fontSize": 16,
                      "fontWeight": "500",
                      "lineHeight": 20,
                    },
                    Object {
                      "color": "#484848",
                    },
                    Object {
                      "fontFamily": "Roboto-Medium",
                      "fontWeight": "500",
                    },
                    Object {
                      "textAlign": "left",
                    },
                  ]
                }
              >
                function mockConstructor() {
        return fn.apply(this, arguments);
      }
              </Text>
              <Text
                style={
                  Array [
                    Array [
                      Object {
                        "color": "#484848",
                        "fontFamily": "Roboto-Light",
                        "fontSize": 13,
                        "lineHeight": 15,
                      },
                      null,
                      null,
                      Object {
                        "textAlign": "left",
                      },
                    ],
                    Object {
                      "marginTop": 4,
                    },
                  ]
                }
              >
                function mockConstructor() {
        return fn.apply(this, arguments);
      }
              </Text>
            </View>
            <View
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <View
                style={
                  Object {
                    "display": "flex",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flexDirection": "row",
                    }
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    autoCorrect={false}
                    autoFocus={true}
                    keyboardType="default"
                    multiline={true}
                    numberOfLines={3}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    rejectResponderTermination={true}
                    spellCheck={false}
                    style={
                      Object {
                        "flex": 3,
                        "fontSize": 20,
                        "padding": 0,
                        "textAlignVertical": "top",
                      }
                    }
                    underlineColorAndroid="transparent"
                    value={null}
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                },
                null,
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                    "flex": 1,
                    "marginBottom": 10,
                  },
                  null,
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#484848",
                      "fontSize": 16,
                      "fontWeight": "500",
                      "lineHeight": 20,
                    },
                    Object {
                      "color": "#484848",
                    },
                    Object {
                      "fontFamily": "Roboto-Medium",
                      "fontWeight": "500",
                    },
                    Object {
                      "textAlign": "left",
                    },
                  ]
                }
              >
                function mockConstructor() {
        return fn.apply(this, arguments);
      }
              </Text>
            </View>
            <View
              style={
                Object {
                  "display": "flex",
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#1877F2",
                    "borderRadius": 4,
                    "display": "flex",
                    "flexDirection": "row",
                    "opacity": 1,
                    "paddingHorizontal": 12,
                    "paddingVertical": 10.5,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Array [
                        Object {
                          "color": "#484848",
                          "fontSize": 16,
                          "fontWeight": "500",
                          "lineHeight": 20,
                        },
                        null,
                        Object {
                          "fontFamily": "Roboto-Bold",
                        },
                        Object {
                          "textAlign": "left",
                        },
                      ],
                      Array [
                        Object {
                          "display": "flex",
                          "flexGrow": 1,
                          "flexShrink": 1,
                        },
                        Object {
                          "color": "white",
                        },
                        Object {
                          "fontSize": 16,
                        },
                        null,
                        Object {
                          "textAlign": "left",
                        },
                      ],
                    ]
                  }
                >
                  <fbt
                    desc="Submit bug button text"
                  >
                    Submit
                  </fbt>
                </Text>
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 20,
                      },
                      null,
                      Object {
                        "fontFamily": "Material Design Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  Û∞Åî
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
